tests/ashley/test_models_lticontext.py

```diff
@@ -0,0 +1,115 @@
+"""
+Tests for the ashley.models.LTIContext model.
+"""
+from django.contrib.auth.models import Group
+from django.test import TestCase
+from machina.core.db.models import get_model
+
+from ashley.factories import LTIContextFactory, UserFactory
+
+GroupForumPermission = get_model(  # pylint: disable=C0103
+    "forum_permission", "GroupForumPermission"
+)
+LTIContext = get_model("ashley", "LTIContext")  # pylint: disable=C0103
+
+
+class LTIContextTestCase(TestCase):
+    """Test the utilities functions of the LTIContext model."""
+
+    def test_group_generators(self):
+        """Test utilities functions to generate django groups related to a LTIContext"""
+
+        context = LTIContextFactory()
+        # Create a django group (corresponding to the "student" role in the LTI context
+        # we just created)
+        initial_student_group = Group.objects.create(
+            name=f"cg__{context.id}__role__student"
+        )
+
+        initial_group_count = Group.objects.count()
+
+        # Ensure that the base group name generator respect a specific name pattern
+        base_group_name = f"cg__{context.id}"
+        self.assertEqual(base_group_name, context.base_group_name)
+
+        base_group = context.base_group()
+        # Ensure that the base group is created if it does not exist
+        self.assertEqual(initial_group_count + 1, Group.objects.count())
+        self.assertEqual(base_group_name, base_group.name)
+
+        # Get the django group corresponding to the "student" role.
+        role_group_student = context.role_group("student")
+        # Ensure that the previously created group is returned and no new group is created
+        self.assertEqual(initial_student_group, role_group_student)
+        self.assertEqual(initial_group_count + 1, Group.objects.count())
+
+        # Get the django group corresponding to the "instructor role
+        role_group_instructor = context.role_group("instructor")
+        # Ensure that the django group is created since it does not exist
+        self.assertEqual(initial_group_count + 2, Group.objects.count())
+        self.assertEqual(
+            f"{base_group_name}__role__instructor", role_group_instructor.name
+        )
+
+    def test_sync_user_groups(self):
+        """Test group synchronization"""
+
+        # Create 2 LTI Contexts
+        context1 = LTIContextFactory()
+        context2 = LTIContextFactory()
+
+        # Create an unrelated django group
+        unrelated_group = Group.objects.create(name="unrelated_django_group")
+
+        # Initialize a user with no group
+        user = UserFactory()
+        self.assertEqual(0, user.groups.count())
+
+        # Sync user groups in context1 with role "student"
+        context1.sync_user_groups(user, ["student"])
+        self.assertCountEqual(
+            [context1.base_group_name, f"{context1.base_group_name}__role__student"],
```

Comment:
why make this to just check the count!?
Why don't you check equality? If it's because you don't know the order, maybe you can check equility of corresponding sets?

Reply:
Despite its misleading name, this is exactly the purpose of [assertCountEqual](https://docs.python.org/3.2/library/unittest.html#unittest.TestCase.assertCountEqual) :

> **assertCountEqual(first, second, msg=None)**
> Test that sequence first contains the same elements as second, regardless of their order. When they don’t, an error message listing the differences between the sequences will be generated.

Reply:
oh wow. I learnt something. shitty name indeed.

Reply:
Yep... I was named `assertSameElements` before and has been deprecated in 3.2 in flavor of this new name. I don't understand why..